internal/client/serverservice.go

```diff
@@ -0,0 +1,202 @@
+	"golang.org/x/sync/semaphore"
+
+	"github.com/metal-toolbox/fleet-scheduler/internal/util"
+	serverserviceapi "go.hollow.sh/serverservice/pkg/api/v1"
+)
+
+const (
+	// server service attribute to look up the BMC IP Address in
+	bmcAttributeNamespace = "sh.hollow.bmc_info"
+)
+
+func newServerserviceClient(ctx context.Context, cfg *util.ServerserviceConfig, logger *logrus.Entry) (*serverserviceapi.Client, error) {
+	if cfg == nil {
+		return nil, ErrNilConfig
+	}
+
+	provider, err := oidc.NewProvider(ctx, cfg.OidcIssuerEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	var clientID string
+	if cfg.OidcClientID != "" {
+		clientID = cfg.OidcClientID
+	} else {
+		clientID = "fleet-scheduler"
+	}
+
+	// setup oauth
+	oauthConfig := clientcredentials.Config{
+		ClientID:       clientID,
+		ClientSecret:   cfg.OidcClientSecret,
+		TokenURL:       provider.Endpoint().TokenURL,
+		Scopes:         cfg.OidcClientScopes,
+		EndpointParams: url.Values{"audience": []string{cfg.OidcAudienceEndpoint}},
+	}
+	oAuthclient := oauthConfig.Client(ctx)
+
+	// set up client
+	retryableClient := retryablehttp.NewClient()
+	retryableClient.HTTPClient = otelhttp.DefaultClient // use otel client so we can collect telemetry
+	retryableClient.HTTPClient.Transport = oAuthclient.Transport
+	retryableClient.HTTPClient.Jar = oAuthclient.Jar
+
+	if logger.Level < logrus.DebugLevel {
+		retryableClient.Logger = nil
+	} else {
+		retryableClient.Logger = logger
+	}
+
+	client := retryableClient.StandardClient()
+	client.Timeout = timeout
+
+	return serverserviceapi.NewClientWithToken(
+		cfg.OidcClientSecret,
+		cfg.Endpoint,
+		client,
+	)
+}
+
+func newServerserviceClientWithoutOAuth(cfg *util.ServerserviceConfig, logger *logrus.Entry) (*serverserviceapi.Client, error) {
+	if cfg == nil {
+		return nil, ErrNilConfig
+	}
+
+	// set up client
+	retryableClient := retryablehttp.NewClient()
+	retryableClient.HTTPClient = otelhttp.DefaultClient // use otel client so we can collect telemetry
+	retryableClient.ResponseLogHook = func(l retryablehttp.Logger, r *http.Response) {
+		if r.StatusCode == http.StatusInternalServerError {
+			b, err := io.ReadAll(r.Body)
+			if err != nil {
+				logger.Warn("serverservice query returned 500 error, got error reading body: ", err.Error())
+				return
+			}
+
+			logger.Warn("serverservice query returned 500 error, body: ", string(b))
+		}
+	}
+
+	if logger.Level < logrus.DebugLevel {
+		retryableClient.Logger = nil
+	} else {
+		retryableClient.Logger = logger
+	}
+
+	client := retryableClient.StandardClient()
+	client.Timeout = timeout
+
+	return serverserviceapi.NewClientWithToken(
+		"dummy",
+		cfg.Endpoint,
+		client,
+	)
+}
+
+func (c *Client) AsyncGatherServers(page_size int) (chan serverserviceapi.Server, *semaphore.Weighted, error) {
+	if c.ssClient == nil {
+		return nil, nil, ErrSsClientIsNil
+	}
+
+	serverCh := make(chan serverserviceapi.Server)
+	concLimiter := semaphore.NewWeighted(int64(page_size*page_size))
```

Comment:
This is replacing the throttle/pauser used here:https://github.com/metal-toolbox/alloy/blob/b42b7599a14967438f5bd73670526efe1bf891b7/internal/collector/collector.go#L384



The goal is to prevent to many calls out to flatten out the curve of calls a bit as we work on them. Most likely this code will never be fully used IMO, but yeah.